iwyu_ast_util.cc

```diff
@@ -946,18 +946,58 @@ bool IsDefaultNewOrDelete(const FunctionDecl* decl,
       !IsBuiltinFile(GetFileEntry(decl)))
     return false;
 
-  const string decl_name = decl->getNameAsString();
```

Comment:
Can you even get rid of the `decl_loc_as_quoted_include` parameter, as well as the code looking at declaration location spelling? Or does `isReplaceableGlobalAllocationFunction` return true for actual operator new overloads in user code as well? It looks like it might.

We want this predicate to only return true if the decl after overload resolution actually refers to a builtin new/delete, right?

I think the code as-is is right, but it would be nice if we _could_ get rid of the string munging on filenames.

Reply:
In general I don't understand how these user-defined overloads work w.r.t. `<new>`: if a user-defined overload uses a weird assortment of expected parameters, how is `<new>` supposed to provide a declaration for it? I looked at `<new>` on my system and it doesn't look like it declares anything for user-defined overloads.

Anyway, I don't understand this particular aspect of C++ (or IWYU) well enough to feel comfortable removing the code you mentioned. But, since you're the expert, I'll defer to your judgment and will remove it if you think it's safe to do so.

Reply:
Off the top of my head, I think it would be the other way around: user-defined overloads need their own declaring header instead of `<new>`.

I realized it didn't really know what was going on here, especially in the light of C++14 and -17 additions, so I read up a bit and ran some experiments, and I think I have a better understanding now, if not entirely complete :)

These two SO answers both make an illuminating argument:
* https://stackoverflow.com/a/6784138/96963
* https://stackoverflow.com/a/9691364/96963
* and an accompanying rant: http://www.scs.stanford.edu/~dm/home/papers/c++-new.html

As I understand it, they says that `operator new` and `new`-expressions are two different things. The former is a function responsible for memory allocation, the latter is a language construct where the compiler generates code to call `operator new`, and then run the constructor on the allocated memory.

Because the language always does both things -- call the allocator, and run the constructor -- a placement-new allocation function becomes necessary, which just roundtrips the pointer that was passed in.

The placement-new (and -delete) functions are _defined_ in `<new>` and thereby aren't recognized by the compiler unless `<new>` is included, and they're also not replaceable. This seems to me to be a poor choice on behalf of the language designers, but I think there's historically been a tendency to avoid magic imbued-by-the-compiler-declarations if a library declaration is possible.

Of course there are subtle exceptions to this rule that placement allocator functions are not replaceable -- they can be overloaded as long as the "placement arguments" are anything other than a single `void*`. Such placement functions can be used (with limited convenience and safety) for arena allocators, etc. It looks like the `std::align_val_t` and `std::nothrow_t` overloads are examples of such special placement functions that happen to have defined semantics.

Anyway, all my research so far indicates that only the "raw" placement-new functions with a `void*` placement parameter are non-replaceable. They also require `<new>` to be included to make that dummy implementation available.

The other placement-like allocation functions (for alignment and nothrow) both use types that are exported from `<new>`, and so IWYU should track these types back to `<new>` without additional handling.

And finally, the remaining allocation functions are somehow magically built into the language, and no header is required. A surprising example is that this compiles without including `<new>`:
```
class Foo {};

int main() {
    // Allocates memory, but does not run a constructor.
    // Surprising because the declaration of void* operator new(size_t) is in <new>.
    Foo* a = (Foo*)::operator new(sizeof(Foo));
    return 0;
}
```

Reply:
I'm going to try and capture this in one or more testcases extracted from badinc.cc, so I can play around with removing the is-declaration-in-header-`<new>` guard clause there.